### PR TITLE
Fix: pixel data mismatch

### PR DESCRIPTION
**Pull request by ducquangkstn**
Sometimes, the VL of pixel data does not match with calculated value. (I can not provide the sample file to test this)
- In this case, we want to return an error rather than continue to parse and returns an EOF error (very hard to debug why)
- This MR also adds a flag to allow ppl to skip the pixelData and mark it as Encapsulated.

### CODE REVIEW

File: element.go [hunk 1]
```diff
@@ -310,8 +310,13 @@ func (e *pixelDataValue) isElementValue()       {}
 func (e *pixelDataValue) ValueType() ValueType  { return PixelData }
 func (e *pixelDataValue) GetValue() interface{} { return e.PixelDataInfo }
 func (e *pixelDataValue) String() string {
-	// TODO: consider adding more sophisticated formatting
-	return ""
+	if len(e.Frames) == 0 {
+		return "empty pixel data"
+	}
+	if e.IsEncapsulated {
+		return fmt.Sprintf("encapsulated FramesLength=%d FrameSize=%d", len(e.Frames), len(e.Frames[0].EncapsulatedData.Data))
```
Comment (suyashkumar) [inline element.go:317]: Perhaps we should indicate that the frame size reported is for the 0th frame?
```suggestion
		return fmt.Sprintf("encapsulated FramesLength=%d Frame[0] size=%d", len(e.Frames), len(e.Frames[0].EncapsulatedData.Data))
```

File: parse.go [hunk 2]
```diff
@@ -74,6 +77,28 @@ func Parse(in io.Reader, bytesToRead int64, frameChan chan *frame.Frame) (Datase
 	return p.dataset, nil
 }
 
+// ParseWithOption parses the entire DICOM at the input io.Reader into a Dataset of DICOM Elements.
+// Allow user to use it with option.
```
Comment (suyashkumar) [inline parse.go:81]: This line can be deleted
```suggestion
```

File: parse.go [hunk 3]
```diff
@@ -74,6 +77,28 @@ func Parse(in io.Reader, bytesToRead int64, frameChan chan *frame.Frame) (Datase
...
+func ParseWithOption(in io.Reader, bytesToRead int64, frameChan chan *frame.Frame, opts ...ParseOption) (Dataset, error) {
```
Comment (suyashkumar) [inline parse.go:82]: I think there is some overlap with #243 here. I think we can consider adding `opts ...ParseOption` to `Parse` directly

File: read.go [hunk 4]
```diff
@@ -110,10 +110,11 @@ func readValue(r dicomio.Reader, t tag.Tag, vr string, vl uint32, isImplicit boo
 	case tag.VRUInt16List, tag.VRUInt32List, tag.VRInt16List, tag.VRInt32List, tag.VRTagList:
 		return readInt(r, t, vr, vl)
 	case tag.VRSequence:
-		return readSequence(r, t, vr, vl)
+		return readSequence(r, t, vr, vl, d)
 	case tag.VRItem:
-		return readSequenceItem(r, t, vr, vl)
+		return readSequenceItem(r, t, vr, vl, d)
 	case tag.VRPixelData:
+
```
Comment (suyashkumar) [inline read.go:117]: nit: rm this newline

File: read.go [hunk 5]
```diff
@@ -269,10 +266,52 @@ func readNativeFrames(d dicomio.Reader, parsedData *Dataset, fc chan<- *frame.Fr
...
+		if err != nil {
+			panic(err)
+		}
+
+		f := frame.Frame{
+			Encapsulated: true,
+			EncapsulatedData: frame.EncapsulatedFrame{
+				Data: data,
+			},
+		}
```
Comment (suyashkumar) [inline read.go:297]: I'm not sure I like this model of parsing is as encapsulated Frame when it isn't actually supposed to be an encapsulated frame. Like trying to use this data as an encapsulated frame wouldn't work as expected later. Maybe we need some kind of error flag to indicate this was a fallback in the frame element.

Can you tell me more about how you need to access this data in your use case? Like if you hit a dicom with this error condition, do you just want to skip doing any downstream work with the pixel data? Or mark the pixel data as unusable or something? Because iiuc in the current model you have no way of distinguishing this case from a case where there was just a valid encapsulated frame parsed (if your option is set to true).

File: read.go [hunk 6]
```diff
@@ -269,10 +266,52 @@ func readNativeFrames(d dicomio.Reader, parsedData *Dataset, fc chan<- *frame.Fr
 
 	debug.Logf("readNativeFrames:\nRows: %d\nCols:%d\nFrames::%d\nBitsAlloc:%d\nSamplesPerPixel:%d", MustGetInts(rows.Value)[0], MustGetInts(cols.Value)[0], nFrames, bitsAllocated, samplesPerPixel)
 
+	bytesAllocated := bitsAllocated / 8
+	bytesRead = bytesAllocated * samplesPerPixel * pixelsPerFrame * nFrames
```
Comment (suyashkumar) [inline read.go:270]: Maybe we should call this variable bytesToRead in this context. By moving it earlier, I do think it changes what it was returning (previously presumably it was returning 0 for a lot of the return statements below), but I don't think this is used much anyway.

File: read.go [hunk 7]
```diff
@@ -269,10 +266,52 @@ func readNativeFrames(d dicomio.Reader, parsedData *Dataset, fc chan<- *frame.Fr
...
+	case uint32(bytesRead) == vl-1 && vl%2 != 0:
+		return nil, 0, fmt.Errorf("vl=%d: %w", vl, ErrorExpectedEvenLength)
+	default:
+		if !parsedData.opts.allowMismatchPixelDataLength {
+			return nil, 0, fmt.Errorf("expected_vl=%d actual_vl=%d %w", bytesRead, vl, ErrorMismatchPixelDataLength)
+		}
+		data := make([]byte, vl)
+		_, err = io.ReadFull(d, data)
+		if err != nil {
+			panic(err)
```
Comment (suyashkumar) [inline read.go:289]: why panic when we can return an error?

File: read_test.go [hunk 8]
```diff
@@ -359,15 +395,15 @@ func TestReadNativeFrames(t *testing.T) {
...
 				mustNewElement(tag.SamplesPerPixel, []int{1}),
 			}},
-			data:              []uint16{1, 2, 3, 4, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+			data:              []uint16{1, 2, 3, 4, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
 			expectedPixelData: nil,
 			expectedError:     ErrorUnsupportedBitsAllocated,
 		},
 		{
-			Name: "3x3, 3 frames, 1 samples/pixel, data bytes with padded 0",
+			Name: "3x3, 3 framesples/pixel, data bytes with padded 0",
```
Comment (suyashkumar) [inline read_test.go:406]: fix the naming please

File: read_test.go [hunk 9]
```diff
@@ -359,15 +395,15 @@ func TestReadNativeFrames(t *testing.T) {
 				mustNewElement(tag.Rows, []int{5}),
 				mustNewElement(tag.Columns, []int{2}),
 				mustNewElement(tag.NumberOfFrames, []string{"1"}),
-				mustNewElement(tag.BitsAllocated, []int{2}),
+				mustNewElement(tag.BitsAllocated, []int{24}),
 				mustNewElement(tag.SamplesPerPixel, []int{1}),
 			}},
-			data:              []uint16{1, 2, 3, 4, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+			data:              []uint16{1, 2, 3, 4, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
```
Comment (suyashkumar) [inline read_test.go:401]: can you explain why these changes were needed please?

File: parse.go [hunk 10]
```diff
@@ -258,16 +283,24 @@ type ParseOption func(*parseOptSet)
...
+func toParseOptSet(opts ...ParseOption) parseOptSet {
+	optSet := parseOptSet{}
 	for _, opt := range opts {
-		opt(optSet)
+		opt(&optSet)
 	}
 	return optSet
 }
 
+// AllowMismatchPixelDataLength allow parser not returning an error when the length calculated from elements do not match with value length
```
Comment (suyashkumar) [inline parse.go:297]: please add a period to the end of the sentence [1], and made some edits as well.

```suggestion
// AllowMismatchPixelDataLength allows the parser to ignore errors when the length calculated from elements do not match with value length.
```

You may also want to explicitly spell out the behavior that the data is still parsed into an errorneous encapsulated pixeldata frame (or remove that behavior if you don't need it in your use case). 

[1] https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

**Review by suyashkumar — COMMENTED** [review]
Hi there, thanks for the PR! Had some comments for you to consider.

Can you also share more about your use case? In particular, if you encounter a dicom like this with this ParseOption enabled, do you intend to use that malformed data in some way? 

Or, if you wish to generally ignore these errors, would it be possible for you to simply inspect this error from Parse and then ignore it in your business logic? Since the main pixeldata should be the last element anyway, you'd still have the whole data set (except in cases of thumbnails [and in those cases you could consider parsing again with an ignore pixel data flag, which is being built in another PR]). 

e.g. 

```
ds, err := Parse(...)
if errors.Is(err, ErrorMismatchPixelDataLength) {
    // Do your custom logic, or ignore, depending on what your usecase needs.
}
```

Anyway, just thought I'd get a better sense of your use case. Thanks!

**Comment by ducquangkstn** [comment]
> Anyway, just thought I'd get a better sense of your use case. Thanks!

We have a DICOM which the PixelData under tag `Icon Image Sequence` is broken.
The main PixelData is fine

```python
import pydicom
dcm = pydicom.dcmread('...')
dcm.pixel_array
array([[1899, 1927, 1875, ..., 4889, 5041, 5266],
       [1906, 1926, 1882, ..., 4865, 5072, 5240],
       [1923, 1907, 1865, ..., 4977, 5050, 5171],
       ...,
       [2288, 2319, 2271, ..., 6861, 6835, 6822],
       [2330, 2358, 2277, ..., 6888, 6835, 6848],
       [2337, 2359, 2301, ..., 6861, 6848, 6749]], dtype=uint16)

dcm.IconImageSequence[0]
(0028, 0002) Samples per Pixel                   US: 1
(0028, 0004) Photometric Interpretation          CS: 'MONOCHROME2'
(0028, 0010) Rows                                US: 64
(0028, 0011) Columns                             US: 64
(0028, 0100) Bits Allocated                      US: 8
(0028, 0101) Bits Stored                         US: 8
(0028, 0102) High Bit                            US: 7
(0028, 0103) Pixel Representation                US: 0
(0028, 1040) Pixel Intensity Relationship        CS: 'LOG'
(7fe0, 0010) Pixel Data                          OB: Array of 1976 elements

dcm.IconImageSequence[0].pixel_array
error
```

My use-case is reading the DICOM and just ignoring the pixelData inside `IconImageSequence`

File: read_test.go (hunk 9 above)
Comment (ducquangkstn) [reply read_test.go:401]: B/c we are moving the checking of `bytesToRead` above, so this change trims the length of data to `bytesToRead`

File: read.go (hunk 5 above)
Comment (ducquangkstn) [reply read.go]: I added another variable at dd6ececbb145ef4c4a4d6baeb9042300823dbd9d (IsVLMismatch).

- Without `AllowMismatchPixelDataLength`, if the VL mismatches calculated `BytesToRead`, the library returns an error.
- With `AllowMismatchPixelDataLength`, the library return a Frame with `VLMismatch=true` indicates that the pixel data is unusable.

File: read.go (hunk 6 above)
Comment (ducquangkstn) [reply read.go]: When we parse the broken DICOM files, the library reads the data in the stream by `bytesToRead` , so it doesn't compare VL to `bytesToRead`  and returns an EOF error, which is not really useful.     

There is some case when the  length remaining bytes is smaller than  `bytesToRead`. So in general, this change improves better error message.

File: read.go [hunk 11]
```diff
@@ -269,10 +265,52 @@ func readNativeFrames(d dicomio.Reader, parsedData *Dataset, fc chan<- *frame.Fr
...
+	}
+
+	skipOne := false
+	switch {
+	case uint32(bytesToRead) == vl: // we good nothing
+	case uint32(bytesToRead) == vl-1 && vl%2 == 0:
+		skipOne = true
+	case uint32(bytesToRead) == vl-1 && vl%2 != 0:
+		return nil, 0, fmt.Errorf("vl=%d: %w", vl, ErrorExpectedEvenLength)
+	default:
```
Comment (suyashkumar) [inline read.go:281]: Looking at this again, I don't love the readability of this block here. I'd like something that makes it clear to the reader that this block is specifically handling the special cases for when `bytesToRead != vl` and for that to be very clear. Needing to have the "do nothing" case further makes this pattern feel weird to me.

Even though this has an additional level of nesting, I wonder if the following is ultimately more readable:

```go
if bytesToRead != vl {
    // Block of code handling all situations where bytesToRead != vl.
    switch {
    bytesToRead == vl-1 && vl%2 == 0:
        skipFinalPaddingByte = true
    bytesToRead == vl-1 && vl%2 !=0:
        return nil, 0, fmt.Errorf("vl=%d: %w", vl, ErrorExpectedEvenLength)
    default:
        // bytesToRead != vl and none of the special conditions are met, so this is an error or fallback behavior should be triggered
        ....
    }
}
```

File: read.go [hunk 12]
```diff
@@ -269,10 +265,52 @@ func readNativeFrames(d dicomio.Reader, parsedData *Dataset, fc chan<- *frame.Fr
...
+			},
+		}
+
+		if fc != nil {
+			fc <- &f
+		}
+		image := PixelDataInfo{
+			IsVLMismatch: true,
+		}
+		image.Frames = append(image.Frames, f)
```
Comment (suyashkumar) [inline read.go:304]: can we move this logic to a helper function, something like `makeErrorPixelData(data io.Reader, frameChannel, errToInsert error) *PixelDataInfo`. This is assuming we go with my frame.ParsingErr suggestion (or potentially ParsingErr could go on the PixelDataInfo outright, since I'm not sure it needs to be frame level). (readNativeFrames is already so complicated, probably best to factor out logic for readability).

File: pkg/frame/frame.go [hunk 13]
```diff
@@ -20,6 +20,8 @@ type CommonFrame interface {
 	GetImage() (image.Image, error)
 	// IsEncapsulated indicates if the underlying Frame is an EncapsulatedFrame.
 	IsEncapsulated() bool
+	// IsVLMismatch indicates if the pixel-data VL mismatches with calculated value
+	IsVLMismatch() bool
```
[the remark this answers is not in the document]
Comment (suyashkumar) [reply pkg/frame/frame.go:24]: and also, would it make sense for this to belong on the PixelDataInfo itself instead of on the frame?

**Review by suyashkumar — COMMENTED** [review]
Thank you! I think this is almost ready, had some comments for your consideration.

File: pkg/frame/frame.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ducquangkstn) [reply pkg/frame/frame.go]: Updated at 95a533fe7792196eb3c1fd61e99336fd5955f4f4

**Comment by ducquangkstn** [comment]
@suyashkumar Pls also approve workflow for enable Github CI for this PR.

File: dataset.go [hunk 14]
```diff
@@ -23,6 +23,8 @@ var ErrorElementNotFound = errors.New("element not found")
 // within this Dataset (including Elements nested within Sequences).
 type Dataset struct {
 	Elements []*Element `json:"elements"`
+
+	opts parseOptSet
```
Comment (suyashkumar) [inline dataset.go:27]: I can take care of this in another PR, but just putting a note for myself: I think if we're going to add this set of options to Dataset, it makes sense that any person who holds a Dataset should be able to inspect the `opts ...ParseOption` the Dataset was parsed with. Typically I don't like just tacking on options to another struct to "pass it through" but in this scenario I think this makes sense because it seems useful for anyone who has a Dataset to see the options it was parsed with.

File: parse.go [hunk 15]
```diff
@@ -50,12 +50,15 @@ var (
 	// has been fully parsed. Users using one of the other Parse APIs should not
 	// need to use this.
 	ErrorEndOfDICOM = errors.New("this indicates to the caller of Next() that the DICOM has been fully parsed")
+
+	// ErrorMismatchPixelDataLength indicates that the size calulated from DICOM mismatch the VL.
+	ErrorMismatchPixelDataLength = errors.New("the size calculated from DICOM elements and VL are mismatch")
```
Comment (suyashkumar) [inline parse.go:55]:
```suggestion
	ErrorMismatchPixelDataLength = errors.New("the size calculated from DICOM elements and the PixelData element's VL are mismatched")
```

File: parse.go [hunk 16]
```diff
@@ -50,12 +50,15 @@ var (
 	// has been fully parsed. Users using one of the other Parse APIs should not
 	// need to use this.
 	ErrorEndOfDICOM = errors.New("this indicates to the caller of Next() that the DICOM has been fully parsed")
+
+	// ErrorMismatchPixelDataLength indicates that the size calulated from DICOM mismatch the VL.
```
Comment (suyashkumar) [inline parse.go:54]:
```suggestion
	// ErrorMismatchPixelDataLength indicates that the size calculated from DICOM mismatches the VL.
```

**Review by suyashkumar — APPROVED** [review]
I think this looks good to me, once the final comments are addressed. I have a couple follow up PRs I may send to build on this as well. Thank you for your contribution!!

File: element.go [hunk 17]
```diff
@@ -296,8 +296,12 @@ func (s *sequencesValue) MarshalJSON() ([]byte, error) {
 
 // PixelDataInfo is a representation of DICOM PixelData.
 type PixelDataInfo struct {
-	Frames         []frame.Frame
-	IsEncapsulated bool `json:"isEncapsulated"`
+	Frames []frame.Frame
+	// ParsingErr indicates if there was an error when reading this Frame from the DICOM.
+	// If this is set, this means fallback behavior was triggered to blindly write the PixelData bytes to an encapsulated frame.
+	// The ParsingErr will contain details about the specific error encountered.
+	ParsingErr     error `json:"parseErr"`
```
[the remark this answers is not in the document]
Comment (suyashkumar) [reply element.go:303]:
```suggestion
	ParseErr     error `json:"parseErr"`
```

[the remark this answers is not in the document]
Comment (ducquangkstn) [reply element.go]: Fixed at dd6ececbb145ef4c4a4d6baeb9042300823dbd9d

File: read.go [hunk 18]
```diff
@@ -221,14 +221,33 @@ func fillBufferSingleBitAllocated(pixelData []int, d dicomio.Reader, bo binary.B
...
 
-// readNativeFrames reads NativeData frames from a Decoder based on already parsed pixel information
-// that should be available in parsedData (elements like NumberOfFrames, rows, columns, etc)
-func readNativeFrames(d dicomio.Reader, parsedData *Dataset, fc chan<- *frame.Frame, vl uint32) (pixelData *PixelDataInfo,
-	bytesRead int, err error) {
+func makeErrorPixelData(reader io.Reader, vl uint32, fc chan<- *frame.Frame, parseErr error) (*PixelDataInfo, error) {
+	data := make([]byte, vl)
+	_, err := io.ReadFull(reader, data)
+	if err != nil {
+		return nil, fmt.Errorf("read pixelData: %w", err)
```
[the remark this answers is not in the document]
Comment (ducquangkstn) [reply read.go]: Fixed at dd6ececbb145ef4c4a4d6baeb9042300823dbd9d

File: parse.go (hunk 16 above)
Comment (ducquangkstn) [reply parse.go]: Fixed at dd6ececbb145ef4c4a4d6baeb9042300823dbd9d

**Comment by suyashkumar** [comment]
🎉 Merged!
